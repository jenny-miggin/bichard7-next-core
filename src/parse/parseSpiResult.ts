import { XMLParser } from "fast-xml-parser"
import { decodeEntitiesProcessor } from "src/lib/encoding"
import type { IncomingMessageParsedXml } from "src/types/SpiResult"
import { incomingMessageParsedXmlSchema } from "src/types/SpiResult"

export default (message: string): IncomingMessageParsedXml => {
  const options = {
    ignoreAttributes: false,
    removeNSPrefix: true,
    parseTagValue: false,
    trimValues: false,
    processEntities: false,
    attributeValueProcessor: decodeEntitiesProcessor,
    tagValueProcessor: decodeEntitiesProcessor
  }

  const parser = new XMLParser(options)
  const rawParsedObj = parser.parse(message) as unknown
  return incomingMessageParsedXmlSchema.parse(rawParsedObj)
}
