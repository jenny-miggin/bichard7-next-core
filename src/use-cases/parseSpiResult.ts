import { XMLParser } from "fast-xml-parser"
import type { IncomingMessageParsedXml } from "src/types/IncomingMessage"
import { incomingMessageParsedXmlSchema } from "src/types/IncomingMessage"

export default (message: string): IncomingMessageParsedXml => {
  const options = {
    ignoreAttributes: false,
    removeNSPrefix: true
  }

  const parser = new XMLParser(options)
  const rawParsedObj = parser.parse(message) as unknown
  return incomingMessageParsedXmlSchema.parse(rawParsedObj)
}
