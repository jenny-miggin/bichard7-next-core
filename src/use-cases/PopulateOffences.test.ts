import { readFileSync } from "fs"
import parseSpiResult from "./parseSpiResult"
import PopulateOffences from "./PopulateOffences"

describe("populateDefendant", () => {
  const message = readFileSync("test-data/input-message-001.xml", "utf-8")
  const courtResult = parseSpiResult(message).DeliverRequest.Message.ResultedCaseMessage

  it("should transform SPI Offences to Hearing Outcome Offences", () => {
    const result = new PopulateOffences(
      courtResult.Session.Case.Defendant.Offence,
      courtResult.Session.CourtHearing.Hearing.DateOfHearing
    ).execute()

    expect(result).toBeDefined()
    expect(result).toMatchSnapshot()
  })
})
