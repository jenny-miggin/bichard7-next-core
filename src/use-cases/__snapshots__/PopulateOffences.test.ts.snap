// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopulateOffences should transform SPI Offences to Hearing Outcome Offences 1`] = `
Object {
  "bailConditions": Array [],
  "offences": Array [
    Object {
      "ActualOffenceDateCode": "1",
      "ActualOffenceEndDate": Object {
        "EndDate": undefined,
      },
      "ActualOffenceStartDate": Object {
        "StartDate": "2010-11-28",
      },
      "ActualOffenceWording": "Attempt to rape a girl aged 13 / 14 / 15 / years of age - SOA 2003.",
      "ArrestDate": "2010-12-01",
      "ChargeDate": "2010-12-02",
      "CommittedOnBail": "D",
      "ConvictionDate": "2011-09-26",
      "CourtOffenceSequenceNumber": "1",
      "CriminalProsecutionReference": Object {
        "OffenceReason": Object {
          "OffenceCode": Object {
            "ActOrSource": "SX",
            "FullCode": "SX03001A",
            "Qualifier": "A",
            "Reason": "001",
            "Year": "03",
          },
        },
      },
      "LocationOfOffence": "Kingston High Street",
      "Result": Array [
        Object {
          "CJSresultCode": "3078",
          "ConvictingCourt": undefined,
          "ModeOfTrialReason": "SUM",
          "NextHearingDate": undefined,
          "PleaStatus": "NG",
          "ResultHearingDate": "2011-09-26",
          "ResultHearingType": "OTHER",
          "ResultQualifierVariable": Array [],
          "ResultVariableText": "Travel Restriction Order",
          "SourceOrganisation": Object {
            "BottomLevelCode": "01",
            "OrganisationUnitCode": "B01EF01",
            "SecondLevelCode": "01",
            "ThirdLevelCode": "EF",
            "TopLevelCode": "B",
          },
          "Verdict": "G",
        },
      ],
    },
    Object {
      "ActualOffenceDateCode": "1",
      "ActualOffenceEndDate": Object {
        "EndDate": undefined,
      },
      "ActualOffenceStartDate": Object {
        "StartDate": "2010-11-28",
      },
      "ActualOffenceWording": "Rape of a Female",
      "ArrestDate": "2010-12-01",
      "ChargeDate": "2010-12-02",
      "CommittedOnBail": "D",
      "ConvictionDate": "2011-09-26",
      "CourtOffenceSequenceNumber": "2",
      "CriminalProsecutionReference": Object {
        "OffenceReason": Object {
          "OffenceCode": Object {
            "ActOrSource": "SX",
            "FullCode": "SX03001",
            "Reason": "001",
            "Year": "03",
          },
        },
      },
      "LocationOfOffence": "Kingston High Street",
      "Result": Array [
        Object {
          "CJSresultCode": "3052",
          "ConvictingCourt": undefined,
          "ModeOfTrialReason": "SUM",
          "NextHearingDate": undefined,
          "PleaStatus": "NG",
          "ResultHearingDate": "2011-09-26",
          "ResultHearingType": "OTHER",
          "ResultQualifierVariable": Array [],
          "ResultVariableText": "defendant must never be allowed out",
          "SourceOrganisation": Object {
            "BottomLevelCode": "01",
            "OrganisationUnitCode": "B01EF01",
            "SecondLevelCode": "01",
            "ThirdLevelCode": "EF",
            "TopLevelCode": "B",
          },
          "Verdict": "G",
        },
      ],
    },
    Object {
      "ActualOffenceDateCode": "1",
      "ActualOffenceEndDate": Object {
        "EndDate": undefined,
      },
      "ActualOffenceStartDate": Object {
        "StartDate": "2010-11-28",
      },
      "ActualOffenceWording": "Use a motor vehicle without third party insurance.",
      "ArrestDate": "2010-12-01",
      "ChargeDate": "2010-12-02",
      "CommittedOnBail": "D",
      "ConvictionDate": "2011-09-26",
      "CourtOffenceSequenceNumber": "3",
      "CriminalProsecutionReference": Object {
        "OffenceReason": Object {
          "OffenceCode": Object {
            "ActOrSource": "RT",
            "FullCode": "RT88191",
            "Reason": "191",
            "Year": "88",
          },
        },
      },
      "LocationOfOffence": "Kingston High Street",
      "Result": Array [
        Object {
          "CJSresultCode": "1015",
          "ConvictingCourt": undefined,
          "ModeOfTrialReason": "SUM",
          "NextHearingDate": undefined,
          "PleaStatus": "NG",
          "ResultHearingDate": "2011-09-26",
          "ResultHearingType": "OTHER",
          "ResultQualifierVariable": Array [],
          "ResultVariableText": "Fined 100.",
          "SourceOrganisation": Object {
            "BottomLevelCode": "01",
            "OrganisationUnitCode": "B01EF01",
            "SecondLevelCode": "01",
            "ThirdLevelCode": "EF",
            "TopLevelCode": "B",
          },
          "Verdict": "G",
        },
      ],
    },
  ],
}
`;
